File: Second-order partial differential equation/OMM_2.m
clear all; clc; close(gcf); 
L=pi; % x in [0, L] 
H=3; % y in [0, H] 
N=101; % 
M=51; % 
T=pi; % 
hx=L/(N-2); % 
hy=H/(M-2); % 
%J=2+round(2*T*sqrt(1/hx^2+1/hy^2)); 
J=101; 
tau=T/(J-1); % 
x=zeros(N,1); % 
y=zeros(M,1); % 
t=zeros(J,1); % 
% --------------------------------------------------- 
for n=1:N 
    x(n)=(n-1)*hx-hx/2; 
end 
for m=1:M 
    y(m)=(m-1)*hy-hy/2; 
end 
for j=1:J 
    t(j)=(j-1)*tau; 
end 
%------------------------------------------------------------------ 
% ------------------------------------------- 
u=zeros(M,N,J); 
for j=1:J 
    for n=1:N 
        for m=1:M 
            u(m,n,j)=(exp(-4*t(j))+(sin(t(j))-cos(t(j))+exp(-4*t(j)))/2)*cos(x(n)); 
        end 
    end 
end 
%------------------------------------------------------------------ 
% ------------------------------------------------ 
v=zeros(M,N,J); % 
err=zeros(M,N,J); % 
% 
for n=1:N 
    for m=1:M 
        v(m,n,1)=cos(x(n)); 
    end 
end
w=zeros(M,N); %
alpha_x=zeros(N-1,1); 
beta_x=zeros(N-1,1); 
alpha_y=zeros(M-1,1); 
beta_y=zeros(M-1,1); 
Ax=1; 
Cx=2*(1+4*hx^2/tau); 
Ay=1; 
Cy=2*(1+4*hy^2/tau); 
for j=1:J-1 
    % -------------- 
    for m=2:M-1 
        alpha_x(1)=1;
        % ---------------------------- 
        for n=2:N-1 
            F=(v(m-1,n,j)-2*v(m,n,j)+v(m+1,n,j))*hx^2/hy^2+2*4*hx^2/tau*v(m,n,j)+4*hx^2*cos(x(n))*sin(t(j)+tau/2); 
            alpha_x(n)=Ax/(Cx-Ax*alpha_x(n-1)); 
            beta_x(n)=(F+Ax*beta_x(n-1))/(Cx-Ax*alpha_x(n-1)); 
        end 
        %------------------------------ 
         w(m,N)=beta_x(N-1)/(1-alpha_x(N-1)); 
        for n=N-1:-1:1 
            w(m,n)=alpha_x(n)*w(m,n+1)+beta_x(n); 
        end 
%         w(m,N)= w(m,N-1);
%         w(m,1) = w(m,2);
    end 
        % --------- 
        alpha_y(1)=1;
    for n=2:N-1 
        % -------------------------- 
        for m=2:M-1 
            F=(w(m,n-1)-2*w(m,n)+w(m,n+1))*hy^2/hx^2+2*4*hy^2/tau*w(m,n)+4*hy^2*cos(x(n))*sin(t(j)+tau/2); 
            alpha_y(m)=Ay/(Cy-Ay*alpha_y(m-1)); 
            beta_y(m)=(F+Ay*beta_y(m-1))/(Cy-Ay*alpha_y(m-1)); 
        end 
        % ------------------------------ 
        v(M,n,j+1)=beta_y(M-1)/(1-alpha_y(M-1)); 
        err(M,n,j+1)=v(M,n,j+1)-u(M,n,j+1); 
        for m=M-1:-1:1 
            v(m,n,j+1)=alpha_y(m)*v(m+1,n,j+1)+beta_y(m); 
            err(m,n,j+1)=v(m,n,j+1)-u(m,n,j+1); 
        end 
    end
    %
    for m=1:M
        v(m,1,j+1) = v(m,2,j+1);
        err(m,1,j+1)=v(m,1,j+1)-u(m,1,j+1); 
        v(m,N,j+1) = v(m,N-1,j+1); 
        err(m,N,j+1)=v(m,N,j+1)-u(m,N,j+1); 
    end 
    for n=1:N
        v(1,n,j+1) = v(2,n,j+1);
        err(1,n,j+1)=v(1,n,j+1)-u(1,n,j+1);
        v(M,n,j+1) = v(M-1,n,j+1); 
        err(M,n,j+1)=v(M,n,j+1)-u(M,n,j+1); 
    end
end
j=20; 
figure 
surf(x,y,u(:,:,j)) 
xlabel('x')
ylabel('y') 
zlabel('u') 
title('analitical solution') 
figure 
surf(x,y,v(:,:,j)) 
xlabel('x') 
ylabel('y') 
zlabel('u') 
title('calculated solution') 
figure 
surf(x,y,err(:,:,j)) 
xlabel('x') 
ylabel('y') 
zlabel('u') 
title('error')

File: Second-order partial differential equation/OMM_2.py
import matplotlib.pyplot as plt
import numpy as np
from tqdm import tqdm


Nx = 160 # число шагов по x
Ny = 160 # число шагов по y
M = 100 #число шагов по t

x = np.linspace(0, 1, Nx)
y = np.linspace(0, 1, Ny)
t = np.linspace(0, 1, M)

h_x = x[1] - x[0]
h_y = y[1] - y[0]
tau = t[1] - t[0]

gamma_x = tau / h_x**2
gamma_y = tau / h_y**2

def lam(k):
    return (np.pi ** 2) * (1 + (2 * k + 1) ** 2)

def exact(x, y, t):
    z = 0
    for k in range(10):
        z = z + (np.sin((np.pi * x) * (2 * k + 1)) * (np.exp(t) - np.exp(-lam (k) * t))) / ((2 * k + 1) * np.pi * (lam(k) + 1))
    return 4 * np.cos(np.pi * y) * z


u = np.zeros((Nx, Ny, 2 * M + 1))

def F_1(i1, i2, j):
    return 0.5 * gamma_y * (u[i1, i2-1, j-1] + u[i1, i2+1, j-1]) + (1 - gamma_y) * u[i1, i2, j-1] + 0.5 * tau * np.exp(tau * (j + 1 / 2)) * np.cos(np.pi * y[i2])

def F_2(i1, i2, j):
    return 0.5 * gamma_x * (u[i1-1, i2, j-1] + u[i1+1, i2, j-1]) + (1 - gamma_x) * u[i1, i2, j-1] + 0.5 * tau * np.exp(tau * (j + 1 / 2)) * np.cos(np.pi * y[i2])

def progonka_x(i2, j):

    d = np.zeros(Nx)
    sigma = np.zeros(Nx)
    d[1] = 0
    sigma[1] = 0
    A = 0.5 * gamma_x
    B = 1 + gamma_x
    C = 0.5 * gamma_x

    for m in range(1, Nx-1):
        Fm = -F_1(m, i2, j)
        d[m+1] = C / (B - A * d[m])
        sigma[m+1] = (Fm - A * sigma[m]) / (A * d[m] - B)
    u[Nx-1, i2, j] = 0
    u[0, i2, j] = 0

    for m in range(Nx-1, 0, -1):
        u[m-1, i2, j] = d[m] * u[m, i2, j] + sigma[m]

def progonka_y(i1, j):

    d = np.zeros(Ny)
    sigma = np.zeros(Ny)
    d[1] = 1
    sigma[1] = 0
    A = 0.5 * gamma_y
    B = 1 + gamma_y
    C = 0.5 * gamma_y
    
    for m in range(1, Ny-1):
        Fm = -F_2(i1, m, j)
        d[m+1] = C / (B - A * d[m])
        sigma[m+1] = (Fm - A * sigma[m]) / (A * d[m] - B)
    u[i1, Ny - 1, j] = sigma[-1] / (1 - d[-1])

    for m in range(Ny-1, 0, -1):
        u[i1, m-1, j] = d[m] * u[i1, m, j] + sigma[m]

u[:, :, 0] = 0 #начальное условие

for j in tqdm(range(1, 2 * M, 2)): #выполнение расчетов
    for i2 in range(1, Ny - 1):
        progonka_x(i2, j)
    for i1 in range(1, Nx - 1):
        progonka_y(i1, j+1)

j_ = 80

p = np.zeros((Nx, Ny))
def p_(j_):
    for i in tqdm(range(Nx)):
        for j in range(Ny):
            p[i][j] = exact(i * h_x, j * h_y, j_ * tau)
    return p

ym = np.linspace(0, 1, num=Ny)
xn = np.linspace(0, 1, num=Nx)
X, Y = np.meshgrid(xn, ym)
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
print(p.shape, X.shape, Y.shape)
nnn = p_(j_)[:, :]
surf_1 = ax.plot_wireframe(X, Y, nnn, rstride=10, cstride=1, color='r')
surf_2 = ax.plot_wireframe(X, Y, u[:, :, j_], rstride=10, cstride=1)
surf_3 = ax.plot_wireframe(X, Y, np.abs(nnn - u[:, :, j_]), rstride=10, cstride=1, color='g')
print(np.mean(np.abs(nnn - u[:, :, j_])))

plt.title('Решение')
plt.xlabel('X')
plt.ylabel('Y')
plt.show()

